crates/noirc_driver/src/lib.rs

```diff
@@ -333,37 +351,45 @@ impl Driver {
         &self,
         options: &CompileOptions,
         main_function: FuncId,
+        np_language: Language,
+        is_opcode_supported: &impl Fn(&Opcode) -> bool,
     ) -> Result<CompiledProgram, FileDiagnostic> {
         let program = monomorphize(main_function, &self.context.def_interner);
 
-        let np_language = self.language.clone();
-
-        let circuit_abi = if options.experimental_ssa {
-            experimental_create_circuit(
-                program,
-                np_language,
-                &self.is_opcode_supported,
-                options.show_ssa,
-                options.show_output,
-            )
+        let (circuit, abi) = if options.experimental_ssa {
+            let (circuit, abi) =
+                experimental_create_circuit(program, options.show_ssa, options.show_output)?;
+
+            let optimized_circuit = {
+                // TODO: Is `abi.field_count` and `circuit.current_witness_index` the same?
+                // If so, the optimization pass can be deduplicated
```

Comment:
```
abi.field_count = How many witnesses are needed to represent the circuit's inputs?
circuit.current_witness_index = How many witnesses are involved in the circuit including intermediate and return values? (actually an upper bound on this number)
```

We shouldn't expect these to be the same except for the simplest circuits.

Reply:
Sounds good. I'll remove the TODO

Reply:
I deduplicated this and used `field_count()` uniformly. Tests still pass and we probably solved a bug we didn't know we had

Reply:
`CircuitSimplifier` is currently a noop as it requires #1333 to do anything.